libs/domain/feature-loaders/src/snapshot/SnapshotProposalLoader.ts

```diff
@@ -0,0 +1,220 @@
+    id: t.string,
+    author: t.string,
+    created: t.number,
+    space: optional(
+        t.strict({
+            id: t.string,
+            name: t.string,
+        })
+    ),
+    type: optional(t.string),
+    strategies: withMessage(
+        t.array(
+            t.strict({
+                name: t.string,
+            })
+        ),
+        () => "strategies failed"
+    ),
+    title: t.string,
+    body: optional(t.string),
+    choices: t.array(t.string),
+    start: t.number,
+    end: t.number,
+    snapshot: t.string,
+    state: t.string,
+    link: optional(t.string),
+    scores: optional(t.array(t.number)),
+    votes: optional(t.number),
+});
+
+export const ProposalsCodec = t.strict({
+    proposals: t.array(ProposalCodec),
+});
+
+export type Proposals = t.TypeOf<typeof ProposalsCodec>;
+
+const INFO = {
+    namespace: "snapshot",
+    name: "Proposal",
+    version: "V1",
+};
+@Nested()
+class Space {
+    @OptionalProperty()
+    name: string;
+    @NonEmptyProperty()
+    id: string;
+}
+
+@Nested()
+class Strategy {
+    @NonEmptyProperty()
+    name: string;
+}
+
+@Data({
+    info: INFO,
+})
+export class Proposal {
+    @NonEmptyProperty()
+    id: string;
+    @NonEmptyProperty()
+    author: string;
+    @NonEmptyProperty()
+    created: number;
+    @OptionalObjectRef(Space)
+    space?: Space;
+    @OptionalProperty()
+    type?: string;
+    @RequiredArrayRef(Strategy)
+    strategies: Strategy[];
+    @NonEmptyProperty()
+    title: string;
+    @OptionalProperty()
+    body?: string;
+    @RequiredStringArrayOf()
+    choices: string[];
+    @NonEmptyProperty()
+    start: number;
+    @NonEmptyProperty()
+    end: number;
+    @NonEmptyProperty()
+    snapshot: string;
+    @NonEmptyProperty()
+    state: string;
+    @OptionalProperty()
+    link?: string;
+    @OptionalNumberArrayOf()
+    scores?: number[];
+    @OptionalProperty()
+    votes: number;
+}
+
+export class SnapshotProposalLoader extends GraphQLDataLoaderBase<
+    Proposals,
+    Proposal
+> {
+    public info = INFO;
+
+    protected cursorMode = "cursor" as const;
+    protected cursor: number;
+
+    protected batchSize = BATCH_SIZE;
+    protected type = Proposal;
+    protected cadenceConfig = {
+        [ScheduleMode.BACKFILL]: { seconds: 5 },
+        [ScheduleMode.INCREMENTAL]: { minutes: 5 },
+    };
+
+    protected graphQLQuery = QUERY;
+    protected codec = ProposalsCodec;
+
+    protected spaces: string[];
+
+    constructor(client: GraphQLClient, spaces: string[]) {
+        super(client);
+        this.spaces = spaces;
+    }
+
+    protected mapResult(result: Proposals): Array<Proposal> {
+        return result.proposals as Array<Proposal>;
+    }
+
+    protected extractCursor = (proposals: Proposals) => {
+        return String(this.cursor + proposals.proposals.length);
+    };
+
+    /**
+     * Safely record the cursor before loading so that we can use it to find the next cursor
+     */
+    protected preLoad = (
```

Comment:
Do we still need this?

Reply:
Yes, for a moment I thought we didn't, but the cursor needs to be saved to the loader at some point before loading.